x/mongo/driver/topology/server.go

```diff
@@ -290,6 +292,32 @@ func (s *Server) ProcessHandshakeError(err error, startingGenerationNumber uint6
 		return
 	}
 
+	isTimeout := func(err error) bool {
+		for err != nil {
+			if netErr, ok := err.(net.Error); ok && netErr.Timeout() {
+				return true
+			}
+			// Handle the case where an error has been replaced by "net.errCanceled", which isn't
+			// exported and can't be compared directly. In this case, just compare the error message.
+			if err.Error() == "operation was canceled" {
+				return true
+			}
+			if wrapper, ok := err.(interface{ Unwrap() error }); ok {
+				err = wrapper.Unwrap()
+			} else {
+				break
+			}
+		}
+
+		return false
+	}
+
+	// Ignore errors that indicate a client-side timeout occurred when using an operation-scoped
+	// deadline (i.e. not using connectTimeoutMS as the connection timeout).
+	if (ctxErr == context.DeadlineExceeded || ctxErr == context.Canceled) && isTimeout(wrappedConnErr) {
```

Comment:
nit: you can check `ctxErr != nil` here as `context.DeadlineExceeded` and `context.Canceled` are the only errors that can be returned from contexts.

Reply:
I considered simplifying to just `ctxErr != nil` but decided to leave it in case `ctx.Err()` is ever updated to return additional errors. After reading the description of `ctx.Err()`, it seems that its behavior is tied directly to `ctx.Done()`, so any new errors would probably indicate some kind of "stop now" condition, which we want to handle in the same way.

Will update to `ctxErr != nil`.